app/lib/activitypub/tag_manager.rb

```diff
@@ -203,6 +203,39 @@ def uri_to_local_id(uri, param = :id)
     path_params[param]
   end
 
+  def uri_to_local_account_params(uri)
+    return unless local_uri?(uri)
+
+    path_params = Rails.application.routes.recognize_path(uri)
+
+    # TODO: handle numeric IDs
+    case path_params[:controller]
+    when 'accounts'
+      [:username, path_params[:username]]
+    when 'instance_actors'
+      [:id, -99]
+    end
+  end
+
+  def uri_to_local_account(uri)
+    param, value = uri_to_local_account_params(uri)
+
+    case param
+    when :username
+      Account.find_local(value)
+    when :id
+      Account.find_by(id: value)
+    end
+  end
+
+  def uris_to_local_accounts(uris)
+    params = uris.filter_map { |uri| uri_to_local_account_params(uri) }
+    usernames = params.filter_map { |param, value| value.downcase if param == :username }
+    ids = params.filter_map { |param, value| value if param == :id }
```

Comment:
I am not a big fan of having three separate array traversals here. I think it should be possible to iterate over `uris` once and gather the `usernames` and `ids` along the way. The resulting code may be a bit more verbose, but might also be clearer and perform better.

Reply:
Yes, I think that is a lot clearer.

If you worry about the verbosity, here are a couple of ideas:

1. Using postfix `if` statements would replace the `case` statement with two lines.
2. This could be a use-case for "pattern matching":
    ```ruby
    case uri_to_local_account_params(uri)
    in [:username, value]
      usernames << value.downcase
    in [:id, value]
      ids << value
    end
    ```
    (Fun fact: I have *never* actually used pattern matching before, mainly because I never encountered - or probably rather never recognized - a proper use-case. This is the first time.)
3. Use `Array#inject` and do something very fancy like `usernames, ids = uris.inject([[], []]) do |memo, param| #...`. But I think that would be a lot less readable.

Reply:
Yeah I suppose the following is ok:
```suggestion
    usernames = []
    ids = []

    uris.each do |uri|
      param, value = instance.send(:uri_to_local_account_params, uri)
      usernames << value.downcase if param == :username
      ids << value if param == :id
    end
```